src/main/java/net/snowflake/ingest/connection/OAuthManager.java

```diff
@@ -0,0 +1,241 @@
+   */
+  OAuthManager(
+      String accountName,
+      String username,
+      OAuthCredential oAuthCredential,
+      URI tokenRequestURI,
+      TelemetryService telemetryService) {
+    this(
+        accountName,
+        username,
+        oAuthCredential,
+        tokenRequestURI,
+        DEFAULT_UPDATE_THRESHOLD_RATIO,
+        telemetryService);
+  }
+
+  /**
+   * Creates a OAuthManager entity for a given account, user and OAuthCredential with a specified
+   * time to renew the token
+   *
+   * @param accountName - the snowflake account name of this user
+   * @param username - the snowflake username of the current user
+   * @param oAuthCredential - the OAuth credential we're using to connect
+   * @param updateThresholdRatio - the ratio between the expiration time of a token and the time
+   *     needed to refresh it.
+   * @param telemetryService reference to the telemetry service
+   */
+  OAuthManager(
+      String accountName,
+      String username,
+      OAuthCredential oAuthCredential,
+      URI tokenRequestURI,
+      double updateThresholdRatio,
+      TelemetryService telemetryService) {
+    super(accountName, username);
+    // if any of our arguments are null, throw an exception
+    if (oAuthCredential == null) {
+      throw new IllegalArgumentException();
+    }
+
+    if (updateThresholdRatio <= 0 || updateThresholdRatio >= 1) {
+      throw new IllegalArgumentException("updateThresholdRation should fall in (0, 1)");
+    }
+
+    this.oAuthCredential = new AtomicReference<>(oAuthCredential);
+
+    this.tokenRequestURI = tokenRequestURI;
+
+    this.httpClient = HttpUtil.getHttpClient(accountName);
+
+    this.telemetryService = telemetryService;
+
+    this.updateThresholdRatio = updateThresholdRatio;
+
+    refreshFailed = new AtomicBoolean();
+
+    // generate our first token
+    refreshToken();
+  }
+
+  @Override
+  String getToken() {
+    if (refreshFailed.get()) {
+      LOGGER.error("getToken request failed due to token refresh failure");
+      throw new SecurityException();
+    }
+    return oAuthCredential.get().getAccessToken();
+  }
+
+  @Override
+  String getTokenType() {
+    return TOKEN_TYPE;
+  }
+
+  /** refreshToken - Get new access token using refresh_token, client_id, client_secret */
+  void refreshToken() {
+    for (int retries = 0; retries < MAX_REFRESH_TOKEN_RETRY; retries++) {
+      String respBodyString = null;
+      try (CloseableHttpResponse httpResponse = httpClient.execute(makeRefreshTokenRequest())) {
+        respBodyString = EntityUtils.toString(httpResponse.getEntity());
+
+        if (httpResponse.getStatusLine().getStatusCode() == HttpStatusCodes.STATUS_CODE_OK) {
+          JsonObject respBody = JsonParser.parseString(respBodyString).getAsJsonObject();
+
+          if (respBody.has(ACCESS_TOKEN) && respBody.has(EXPIRES_IN)) {
+            String newAccessToken = respBody.get(ACCESS_TOKEN).toString().replaceAll("^\"|\"$", "");
```

Comment:
What is this replace targeting?

Reply:
The parsed result would be "<token>", this regex is trimming the start and end quotation.